File: web/src/products/curation/domain/thread.ts
import type { BudgetLedger } from "./budget";
export type ControlMode = { mode: "AUTO" | "MANUAL"; version: number };
export type ActionEffect = { kind: string; targetId?: string; targetLabel?: string; before?: BudgetLedger | unknown; after?: BudgetLedger | unknown; count?: number };
// Round outcome a completed research Job reports (ADR-0086). Every number a response shows comes from here.
export type ResearchSourceFact = { source: string; status: string; reasonCode?: string; candidateCount: number };
export type ResearchFacts = { roundId: string; observed: number; duplicates: number; rejected: number; admitted: number; evaluated: number; unevaluated: number; sources: ResearchSourceFact[] };
export type ActionJob = { jobId: string; attemptId?: string; actionId: string; kind: string; targetId?: string; targetLabel?: string; status: string; effects: ActionEffect[]; reasonCode?: string; facts?: ResearchFacts };
// The Thread's natural-language reply. `[[ref]]` tokens in the body resolve through `references`, so a product name is never the model's own spelling.
export type ResponseReference = { ref: string; candidateId: string; targetId: string; title: string };
export type ThreadResponse = { schemaVersion: "vitlane.thread-response.v1" | "vitlane.thread-response.v2"; combination?: import("./combination").Combination; kind: "COMMENT" | "ANSWER"; body: string; locale: string; references: ResponseReference[]; modelKey?: string; createdAt: string };
export type ActionDecision = { id: string; kind: "ACTION" | "TARGET" | "CONDITIONS" | "BUDGET"; result: string; targetId?: string; targetLabel?: string; source: string; evidence?: string; reasonCode: string; actionIds: string[] };
export type ActionQuestion = { id: string; prompt: string; options: { id: string; label: string }[] };
export type CurationActionExecution = { id: string; threadId: string; sequence: number; type: string; targetId?: string; targetLabel?: string; status: "PENDING" | "RUNNING" | "WAITING_SELECTION" | "SUCCEEDED" | "FAILED" | "CANCELLED" | "SKIPPED"; generatedByActionId?: string; decisionIds: string[]; retryOfActionId?: string; jobs: ActionJob[]; effects: ActionEffect[]; decisions: ActionDecision[]; question?: ActionQuestion; questions?: ActionQuestion[]; answers: { revision: number; questionId: string; optionId?: string; text?: string }[]; reasonCode?: string; instruction?: string; response?: ThreadResponse };
export type CurationThread = { schemaVersion: "vitlane.curation-thread.v2"; id: string; curationId: string; retryOfThreadId?: string; retryOfJobId?: string; targetLabels?: Record<string, string>; mode: "AUTO" | "MANUAL"; origin: string; request: string; revision: number; status: string; actions: CurationActionExecution[]; reasonCode?: string; createdAt: string; updatedAt: string };
export const threadActive = (t: CurationThread) => ["INTERPRETING", "WAITING_SELECTION", "RUNNING"].includes(t.status);
export const currentAction = (t: CurationThread) => t.actions.find(a => ["PENDING", "RUNNING", "WAITING_SELECTION"].includes(a.status));

// The Server keeps the Thread status enum closed and reports a partial failure
// (some parallel Jobs succeeded, later Actions stopped) through this reason.
export const PARTIAL_FAILURE = "PARTIAL_FAILURE";
export type ThreadOutcome = "ACTIVE" | "SUCCEEDED" | "NO_RESULTS" | "SETTINGS_ONLY" | "ANSWERED" | "PARTIAL" | "FAILED" | "CANCELLED";
const researchActionTypes = ["START_RESEARCH", "PLANNING_START_CURATING", "TARGET_RESEARCH_AGAIN", "CURATION_ADD_TARGETS", "PLANNING_ADD_TARGETS", "INTENT_NEXT_STEP"];
export const threadJobs = (t: CurationThread): ActionJob[] => t.actions.flatMap(a => a.jobs);
export const researchJobs = (t: CurationThread) => threadJobs(t).filter(j => j.kind === "RESEARCH_ROUND");
export const failedJobs = (t: CurationThread) => threadJobs(t).filter(j => j.status === "FAILED");
export const candidateCount = (j: ActionJob) => j.effects.filter(e => e.kind === "CANDIDATES_ADDED").reduce((sum, e) => sum + (e.count ?? 0), 0);
export const targetLabelOf = (t: CurationThread, targetId?: string, fallback = "") => (targetId && t.targetLabels?.[targetId]) || fallback;
export function threadOutcome(t: CurationThread): ThreadOutcome {
 if (threadActive(t)) return "ACTIVE";
 if (t.status === "CANCELLED") return "CANCELLED";
 if (t.status === "FAILED") return t.reasonCode === PARTIAL_FAILURE ? "PARTIAL" : "FAILED";
 const research = researchJobs(t).filter(j => j.status === "SUCCEEDED");
 // A question ran no primitive: its only work is the reply, written or not.
 if (research.length === 0 && responseAction(t)?.instruction === "ANSWER") return "ANSWERED";
 if (research.length === 0) return t.actions.some(a => researchActionTypes.includes(a.type)) ? "SUCCEEDED" : "SETTINGS_ONLY";
 return research.every(j => candidateCount(j) === 0) ? "NO_RESULTS" : "SUCCEEDED";
}
// Candidates found per target across every completed research Job of the Thread.
export function candidateSummary(t: CurationThread, fallbackLabel: string) {
 const byTarget = new Map<string, { targetId: string; label: string; count: number }>();
 for (const j of researchJobs(t)) {
  if (j.status !== "SUCCEEDED") continue;
  const key = j.targetId ?? j.jobId;
  const entry = byTarget.get(key) ?? { targetId: key, label: j.targetLabel || targetLabelOf(t, j.targetId, fallbackLabel), count: 0 };
  entry.count += candidateCount(j);
  byTarget.set(key, entry);
 }
 return [...byTarget.values()];
}
export const RESPONSE_ACTION = "RESPONSE";
export const responseAction = (t: CurationThread) => t.actions.find(a => a.type === RESPONSE_ACTION);
export const threadResponse = (t: CurationThread) => responseAction(t)?.response;
export const threadProgress = (t: CurationThread) => ({ done: t.actions.filter(a => a.status === "SUCCEEDED").length, total: t.actions.length });

File: server/internal/curation/domain/thread.go
package domain

import (
	"encoding/json"
	"github.com/vitlane/vitlane/server/internal/shared/fault"
	"time"
)

const ThreadSchema = "vitlane.curation-thread.v2"

type ControlMode struct {
	Mode    string `json:"mode"`
	Version int64  `json:"version"`
}
type ActionDecision struct {
	ID          string   `json:"id"`
	Kind        string   `json:"kind"`
	Result      string   `json:"result"`
	TargetID    string   `json:"targetId,omitempty"`
	TargetLabel string   `json:"targetLabel,omitempty"`
	Source      string   `json:"source"`
	Evidence    string   `json:"evidence,omitempty"`
	ReasonCode  string   `json:"reasonCode"`
	ActionIDs   []string `json:"actionIds"`
}
type ActionEffect struct {
	Kind        string          `json:"kind"`
	TargetID    string          `json:"targetId,omitempty"`
	TargetLabel string          `json:"targetLabel,omitempty"`
	Before      json.RawMessage `json:"before,omitempty"`
	After       json.RawMessage `json:"after,omitempty"`
	Count       int             `json:"count,omitempty"`
}
type ActionJobResult struct {
	AttemptID  string         `json:"attemptId,omitempty"`
	JobID      string         `json:"jobId"`
	ActionID   string         `json:"actionId"`
	Kind       string         `json:"kind"`
	TargetID   string         `json:"targetId,omitempty"`
	Status     string         `json:"status"`
	Effects    []ActionEffect `json:"effects"`
	ReasonCode string         `json:"reasonCode,omitempty"`
	// Facts is the Round outcome of a completed research Job.
	Facts *ResearchFacts `json:"facts,omitempty"`
}
type ThreadOption struct {
	ID        string           `json:"id"`
	Label     string           `json:"label"`
	TargetID  string           `json:"targetId,omitempty"`
	Actions   []CurationAction `json:"actions,omitempty"`
	Decisions []ActionDecision `json:"decisions"`
}
type ThreadQuestion struct {
	ID      string         `json:"id"`
	Prompt  string         `json:"prompt"`
	Options []ThreadOption `json:"options"`
}
type ThreadAnswer struct {
	Revision   int64  `json:"revision"`
	QuestionID string `json:"questionId"`
	OptionID   string `json:"optionId,omitempty"`
	Text       string `json:"text,omitempty"`
}
type CurationThread struct {
	ExpectedConversationVersion *int64            `json:"expectedConversationVersion,omitempty"`
	RetryOfThreadID             string            `json:"retryOfThreadId,omitempty"`
	RetryOfJobID                string            `json:"retryOfJobId,omitempty"`
	SchemaVersion               string            `json:"schemaVersion"`
	TargetLabels                map[string]string `json:"targetLabels"`
	ID                          string            `json:"id"`
	UserID                      string            `json:"-"`
	CurationID                  string            `json:"curationId"`
	PlanID                      string            `json:"-"`
	AuthSessionID               string            `json:"-"`
	Mode                        string            `json:"mode"`
	Origin                      string            `json:"origin"`
	Request                     string            `json:"request"`
	RequestHash                 string            `json:"-"`
	ExpectedCurationVersion     int64             `json:"expectedCurationVersion"`
	Revision                    int64             `json:"revision"`
	Status                      string            `json:"status"`
	Actions                     []CurationAction  `json:"actions"`
	ReasonCode                  string            `json:"reasonCode,omitempty"`
	CreatedAt                   time.Time         `json:"createdAt"`
	UpdatedAt                   time.Time         `json:"updatedAt"`
}

func (t CurationThread) Active() bool {
	return t.Status == "INTERPRETING" || t.Status == "WAITING_SELECTION" || t.Status == "RUNNING"
}
func (t *CurationThread) CurrentAction() *CurationAction {
	for i := range t.Actions {
		if !t.Actions[i].Terminal() {
			return &t.Actions[i]
		}
	}
	return nil
}
func (t *CurationThread) Cancel() {
	t.Status = "CANCELLED"
	t.Revision++
	for i := range t.Actions {
		a := &t.Actions[i]
		if a.Status == "PENDING" {
			a.Status = "SKIPPED"
		} else if !a.Terminal() {
			a.Status = "CANCELLED"
			a.Question = nil
		}
	}
}
func (t *CurationThread) RefreshStatus() {
	if !t.Active() {
		return
	}
	for i := range t.Actions {
		a := &t.Actions[i]
		if a.Status == "FAILED" || a.Status == "CANCELLED" {
			t.Status = a.Status
			t.ReasonCode = a.ReasonCode
			for j := i + 1; j < len(t.Actions); j++ {
				if !t.Actions[j].Terminal() {
					t.Actions[j].Status = "SKIPPED"
				}
			}
			return
		}
		if a.Terminal() {
			continue
		}
		t.Status = "RUNNING"
		if a.Status == "WAITING_SELECTION" {
			t.Status = "WAITING_SELECTION"
		} else if a.Type == CurationActionAutoStart {
			t.Status = "INTERPRETING"
		}
		return
	}
	t.Status = "SUCCEEDED"
}

type ActionPlan struct {
	Actions   []CurationAction
	Decisions []ActionDecision
	Question  *ThreadQuestion
}

func (t ActionPlan) ValidatePlan() error {
	invalid := func() error { return fault.New(fault.ProviderRejected, "AUTO_PLAN_INVALID", false) }
	if len(t.Actions) > 24 || len(t.Decisions) > 64 {
		return invalid()
	}
	if t.Question != nil {
		if len(t.Actions) != 0 || len(t.Question.Options) < 2 || len(t.Question.Options) > 3 || t.Question.Prompt == "" {
			return invalid()
		}
		for _, option := range t.Question.Options {
			if option.Label == "" {
				return invalid()
			}
			if err := (ActionPlan{Actions: option.Actions, Decisions: option.Decisions}).ValidatePlan(); err != nil {
				return err
			}
		}
	}
	for _, s := range t.Actions {
		switch s.Type {
		case "BUDGET_CHANGE":
			if s.Budget == nil {
				return invalid()
			}
		case "CRITERIA_CHANGE":
			if s.Criteria == nil || s.Criteria.Validate() != nil || s.TargetID == "" {
				return invalid()
			}
		case "TARGET_RESEARCH_AGAIN":
			if s.TargetID == "" {
				return invalid()
			}
		case "CURATION_ADD_TARGETS", "PLANNING_ADD_TARGETS":
			if s.Instruction == "" {
				return invalid()
			}
		case CurationActionResponse:
			// A planned RESPONSE is the answer to a question and runs alone.
			if (s.Instruction != ResponseKindAnswer && s.Instruction != "COMBINATION") || len(t.Actions) != 1 {
				return invalid()
			}
		default:
			return invalid()
		}
	}
	for _, d := range t.Decisions {
		if d.Kind != "ACTION" && d.Kind != "TARGET" && d.Kind != "CONDITIONS" && d.Kind != "BUDGET" {
			return invalid()
		}
	}
	if t.Question == nil && len(t.Actions) == 0 {
		return invalid()
	}
	return nil
}
